cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,266 @@
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Loading source index")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	debug.Log("Loading destination index")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	dstSnapshotByOriginal := make(map[restic.ID][]*restic.Snapshot)
+	for sn := range FindFilteredSnapshots(ctx, dstRepo, opts.Hosts, opts.Tags, opts.Paths, nil) {
+		if sn.Original != nil && !sn.Original.IsNull() {
+			dstSnapshotByOriginal[*sn.Original] = append(dstSnapshotByOriginal[*sn.Original], sn)
+		}
+		// also consider identical snapshot copies
+		dstSnapshotByOriginal[*sn.ID()] = append(dstSnapshotByOriginal[*sn.ID()], sn)
+	}
+
+	cloner := &treeCloner{
+		srcRepo:      srcRepo,
+		dstRepo:      dstRepo,
+		visitedTrees: restic.NewIDSet(),
+		buf:          nil,
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Hosts, opts.Tags, opts.Paths, args) {
+		Verbosef("\nsnapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+
+		// check whether the destination has a snapshot with the same persistent ID which has similar snapshot fields
+		srcOriginal := *sn.ID()
+		if sn.Original != nil {
+			srcOriginal = *sn.Original
+		}
+		if originalSns, ok := dstSnapshotByOriginal[srcOriginal]; ok {
+			isCopy := false
+			for _, originalSn := range originalSns {
+				if similarSnapshots(originalSn, sn) {
+					Verbosef("skipping source snapshot %s, was already copied to snapshot %s\n", sn.ID().Str(), originalSn.ID().Str())
+					isCopy = true
+					break
+				}
+			}
+			if isCopy {
+				continue
+			}
+		}
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := cloner.copyTree(ctx, *sn.Tree); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs and saved index")
+
+		// save snapshot
+		sn.Parent = nil // Parent does not have relevance in the new repo.
+		// Use Original as a persistent snapshot ID
+		if sn.Original == nil {
+			sn.Original = sn.ID()
+		}
+		newID, err := dstRepo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		Verbosef("snapshot %s saved\n", newID.Str())
+	}
+	return nil
+}
+
+func similarSnapshots(sna *restic.Snapshot, snb *restic.Snapshot) bool {
+	// everything except Parent and Original must match
+	if !sna.Time.Equal(snb.Time) || !sna.Tree.Equal(*snb.Tree) || sna.Hostname != snb.Hostname ||
+		sna.Username != snb.Username || sna.UID != snb.UID || sna.GID != snb.GID ||
+		len(sna.Paths) != len(snb.Paths) || len(sna.Excludes) != len(snb.Excludes) ||
+		len(sna.Tags) != len(snb.Tags) {
+		return false
+	}
+	for i, a := range sna.Tags {
+		if a != snb.Tags[i] {
+			return false
+		}
+	}
+	for i, a := range sna.Paths {
+		if a != snb.Paths[i] {
+			return false
+		}
+	}
+	for i, a := range sna.Excludes {
+		if a != snb.Excludes[i] {
+			return false
+		}
```

Comment:
IIRC Tags and Paths (and maybe Excludes) are not sorted, so this might miss some "similar" snapshots. However, as we are only comparing snapshots where one should be anyway the original snapshot of another this is ok, IMO.

Reply:
I've changed the check for the Tags and Paths to ignore the ordering, as is done by other commands which filter snapshots.